students/MichaelM/lesson7/html_render.py

```diff
@@ -1,52 +1,124 @@
+                file_out.write(mc)
+            file_out.write('\n')
+            counter = counter + 1
+        file_out.write(cur_ind + f"</{self.tag}>")
+
+
+def render_page(HtmlPage, filename, indent=None):
+    f = StringIO() # StringIO object (like a file, but in memory) to render to memory
+    if indent is None:
+        HtmlPage.render(f)
+    else:
+        HtmlPage.render(f, indent)
+    print(f.getvalue())
+    with open(filename, 'w') as outfile:
+        outfile.write(f.getvalue())
+
+
+class A(Element):
+    tag = 'a'
 
-    def append_to_list(self, appended_txt):
-        self.content_list.append(appended_txt)
+    def __init__(self, link, content):
+        super().__init__(content, href=link)
+
+
+class Html(Element):
+    tag="html"
 
     def render(self, file_out, cur_ind=""):
-        file_out.write(
-            f"{cur_ind}".join(self.content_list))
-
-    def html(self, tag_type, cur_ind=""):
-        if tag_type == "open":
-            self.append_to_list(f"{cur_ind}<html>\n")
-        elif tag_type == "close":
-            self.append_to_list(f"{cur_ind}</html>\n")
-
-    def head(self, tag_type, cur_ind="", att_list=[]):
-        if tag_type == "open":
-            self.append_to_list(f"{cur_ind}<head>\n")
-        elif tag_type == "close":
-            self.append_to_list(f"{cur_ind}</head>\n")
-
-    def title(self, title_txt="", cur_ind=""):
-        self.append_to_list(f"{cur_ind}<title>{title_txt}</title>\n")
-
-    def body(self, tag_type, cur_ind="", att_list=[]):
-        if tag_type == "open":
-            self.append_to_list(f"{cur_ind}<body>\n")
-        elif tag_type == "close":
-            self.append_to_list(f"{cur_ind}</body>\n")
-
-    def br(self, cur_ind=""):
-        self.append_to_list(f"{cur_ind}</br>\n")
-
-    def p(self, tag_type, cur_ind="", att_list=[]):
-        if tag_type == "open":
-            self.append_to_list(f"{cur_ind}<p>")
-        elif tag_type == "close":
-            self.append_to_list(f"{cur_ind}</p>\n")
+        file_out.write("<!DOCTYPE html>\n")
+        super().render(file_out, cur_ind)
+
+
+class Body(Element):
+    tag="body"
+
+
+class P(Element):
+    tag="p"
+
+
+class Head(Element):
+    tag = 'head'
+
+
+class OneLineTag(Element):
+    def __init__(self, content=""):
+        self.content = content
+
+    def render(self, file_out, cur_ind=""):
+        file_out.write(cur_ind + f"<{self.tag}> {self.content} </{self.tag}>")
+
+
+class H(OneLineTag):
+    def __init__(self, num ,content):
+        self.tag = 'h' + str(num)
+        OneLineTag.__init__(self, content)
+
+class Title(OneLineTag):
+    tag = 'title'
+
+
+class Singleton(Element):
+    def render(self, file_out, cur_ind=""):
+        file_out.write(cur_ind + f"<{self.tag} />")
+
+
+class SelfClosingTag(Element):
+    def __init__(self, content="", **kwargs):
+        self.content = content
+        self.kwargs = kwargs
+
+    def render(self, file_out, cur_ind=''):
+        if self.kwargs == {}:
```

Comment:
you technically don't need this check - your code will work the same with just the lines 102, 103